Mathlib/Algebra/Homology/EulerPoincare.lean

```diff
@@ -0,0 +1,319 @@
+          rw [dim_image]
+    _ = (Module.finrank k (LinearMap.ker (C.dFrom i).hom) : ℤ) := by
+
+          have h := Submodule.finrank_quotient_add_finrank
+            (LinearMap.range S.moduleCatToCycles : Submodule k (LinearMap.ker (C.dFrom i).hom))
+
+          have eq1 : Module.finrank k data.H =
+            Module.finrank k (LinearMap.ker (C.dFrom i).hom ⧸
+              LinearMap.range S.moduleCatToCycles) := by
+            rw [data_H_eq]
+          rw [eq1]
+          norm_cast
+
+/-- Factoring out -1 from an alternating sum shifts the sign pattern:
+    ∑ i ∈ range n, (-1)^(i+1) * f i = -∑ i ∈ range n, (-1)^i * f i -/
+private lemma sum_range_alternating_pow_neg_one_succ {α : Type*} [Ring α] (n : ℕ) (f : ℕ → α) :
+    ∑ i ∈ Finset.range n, (-1 : α)^(i + 1) * f i = -∑ i ∈ Finset.range n, (-1 : α)^i * f i := by
+  simp_rw [pow_succ]
+  rw [← Finset.sum_neg_distrib]
+  apply Finset.sum_congr rfl
+  intros x _
+  simp only [neg_mul, mul_neg, mul_one]
+
+/-- Alternating sum lemma with shift decomposition for the Euler-Poincaré formula.
+    If a sequence `a` decomposes as `a(i) = h(i) + b(i) + c(i)` where the terms
+    satisfy a shift relation `b(i+1) = c(i)` with boundary conditions `b(0) = 0`
+    and `c(n) = 0`, then the alternating sum of `a` equals the alternating sum of `h`
+    (the b and c terms telescope and cancel). -/
+private lemma sum_range_alternating_shift_decomp {α : Type*} [Ring α] (n : ℕ)
+    (a b c h : ℕ → α)
+    (h_decomp : ∀ i ≤ n, a i = h i + b i + c i)
+    (h_shift : ∀ i < n, b (i + 1) = c i)
+    (h_b0 : b 0 = 0)
+    (h_cn : c n = 0) :
+    ∑ i ∈ Finset.range (n + 1), (-1 : α)^i * a i =
+    ∑ i ∈ Finset.range (n + 1), (-1 : α)^i * h i := by
+  -- Expand using decomposition
+  have : ∑ i ∈ Finset.range (n + 1), (-1 : α)^i * a i =
+      ∑ i ∈ Finset.range (n + 1), (-1 : α)^i * (h i + b i + c i) := by
+    apply Finset.sum_congr rfl
+    intros i hi
+    simp only [Finset.mem_range] at hi
+    rw [h_decomp i (Nat.lt_succ_iff.mp hi)]
+  rw [this]
+  simp_rw [mul_add]
+  rw [Finset.sum_add_distrib, Finset.sum_add_distrib]
+
+  -- Show b and c terms cancel
+  suffices h_cancel : ∑ i ∈ Finset.range (n + 1), (-1 : α)^i * b i +
+                     ∑ i ∈ Finset.range (n + 1), (-1 : α)^i * c i = 0 by
+    rw [add_assoc, h_cancel, add_zero]
+
+  -- Handle c sum
+  have hc : ∑ i ∈ Finset.range (n + 1), (-1 : α)^i * c i =
+      ∑ i ∈ Finset.range n, (-1 : α)^i * b (i + 1) := by
+    rw [Finset.sum_range_succ]
+    simp [h_cn]
+    apply Finset.sum_congr rfl
+    intros i hi
+    simp only [Finset.mem_range] at hi
+    congr 1
+    exact (h_shift i hi).symm
+  rw [hc]
+
+  -- Handle b sum
+  have hb : ∑ i ∈ Finset.range (n + 1), (-1 : α)^i * b i =
+      ∑ j ∈ Finset.range n, (-1 : α)^(j + 1) * b (j + 1) := by
+    rw [Finset.sum_range_succ']
+    simp [h_b0, pow_zero]
+  rw [hb]
+  rw [sum_range_alternating_pow_neg_one_succ]
+  simp
+
+set_option maxHeartbeats 400000 in -- need to increase heartbeats for final simp
```

Comment:
Can you `set_option diagnostics true` here and see what `simp` is doing? This seems bad.

Reply:
Looks like there's quite a lot of unfolding of instances and reducible definitions, especially in `ModuleCat.carrier`. I tweaked things a bit by using a simple `rw` and then `rfl`. Phew!